File: minimum-window-substring/Solution76.java
// https://leetcode.com/problems/minimum-window-substring/

import java.util.HashMap;

class Solution76 {

    // # T: O(n), S: O(k)
    public String minWindow(String s, String t) {
        if (s == null || t == null || s.length() == 0 || t.length() == 0) {
            return "";
        }

        var tCounts = new HashMap<Character, Integer>();
        for (char ch : t.toCharArray()) {
            tCounts.put(ch, tCounts.getOrDefault(ch, 0) + 1);
        }
        int required = tCounts.size();

        int minLeft = 0, minRight = 0, minWinLen = Integer.MAX_VALUE;

        // Keep track of how many unique characters in t are present
        // in the current window in its desired frequency
        int formed = 0;

        var winCounts = new HashMap<Character, Integer>();

        for (int left = 0, right = 0; right < s.length(); right++) {
            char rightChar = s.charAt(right);
            winCounts.put(rightChar, winCounts.getOrDefault(rightChar, 0) + 1);

            if (tCounts.containsKey(rightChar)
                    && winCounts.get(rightChar).intValue() == tCounts.get(rightChar).intValue()) {
                formed++;
            }
            // Shrink the window till the point where it ceases to be formed
            while (left <= right && formed == required) {
                int currWinLen = right - left + 1;
                if (currWinLen < minWinLen) {
                    minWinLen = currWinLen;
                    minLeft = left;
                    minRight = right;
                }
                // Left char no longer part of the window
                char leftChar = s.charAt(left);
                winCounts.put(leftChar, winCounts.get(leftChar) - 1);
                // If the frequency of the current character becomes less than the desired count
                // decrement formed
                if (tCounts.containsKey(leftChar)
                        && winCounts.get(leftChar).intValue() < tCounts.get(leftChar).intValue()) {
                    formed--;
                }
                left++;
            }
        }

        return minWinLen == Integer.MAX_VALUE ? "" : s.substring(minLeft, minRight + 1);
    }

    public static void main(String[] args) {
        var sol = new Solution76();
        var s1 = "ADOBECODEBANC";
        var t1 = "ABC";
        System.out.println(sol.minWindow(s1, t1)); // "BANC"
        var s2 = "a";
        var t2 = "a";
        System.out.println(sol.minWindow(s2, t2)); // "a"
        var s3 = "a";
        var t3 = "aa";
        System.out.println(sol.minWindow(s3, t3)); // ""
    }

}

File: minimum-window-substring/solution76.py
# https://leetcode.com/problems/minimum-window-substring/

class Solution76:

    # T: O(n), S: O(k)
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t:
            return ""

        t_counts = {}
        for ch in t:
            t_counts[ch] = t_counts.get(ch, 0) + 1
        required = len(t_counts)

        min_left, min_right, min_win_len = 0, 0, float('inf')
        formed = 0
        win_counts = {}

        left = 0
        for right, right_char in enumerate(s):
            win_counts[right_char] = win_counts.get(right_char, 0) + 1

            if right_char in t_counts and win_counts[right_char] == t_counts[right_char]:
                formed += 1

            while left <= right and formed == required:
                curr_win_len = right - left + 1
                if curr_win_len < min_win_len:
                    min_win_len = curr_win_len
                    min_left = left
                    min_right = right

                left_char = s[left]
                win_counts[left_char] -= 1
                if left_char in t_counts and win_counts[left_char] < t_counts[left_char]:
                    formed -= 1
                left += 1

        return "" if min_win_len == float('inf') else s[min_left:min_right + 1]


def main() -> None:
    sol = Solution76()
    s1 = "ADOBECODEBANC"
    t1 = "ABC"
    print(sol.minWindow(s1, t1))  # "BANC"
    s2 = "a"
    t2 = "a"
    print(sol.minWindow(s2, t2))  # "a"
    s3 = "a"
    t3 = "aa"
    print(sol.minWindow(s3, t3))  # ""


if __name__ == "__main__":
    main()
